engine/engine-paper/src/main/kotlin/com/typewritermc/engine/paper/interaction/ChatHistoryHandler.kt

```diff
@@ -56,12 +103,19 @@ class ChatHistoryHandler :
             val component = message.message
             val history = getHistory(event.user.uuid)
             if (component is TextComponent && component.content().startsWith("no-index")) {
-                if (component.content().endsWith("resend")) return
-
                 history.allowedMessageThrough()
                 return
             }
+
+            val player = server.getPlayer(event.user.uuid) ?: return
+            if (ResendTokenRegistry.consume(player, component)) {
+                return
+            }
+
             if (component.shouldSaveMessage()) {
+                if (!history.shouldBlockMessage() && message is Message.PlayerMessage) {
+                    message.packet = null
```

Comment:
I don't think this is the appropriate location todo this. I think the message class should have a clear operation or something. And this logic would be more appropriate to be called in the add message function in the chat history class